// clipper2-ts/src/index.ts
/*******************************************************************************
* Author    :  Angus Johnson 
* Date      :  2025
* Website   :  https://www.angusj.com
* Copyright :  Angus Johnson 2010-2025
* License   :  https://www.boost.org/LICENSE_1_0.txt
*******************************************************************************/

// Export core types and interfaces
export {
  type Point64,
  type PointD,
  type Path64,
  type PathD,
  type Paths64,
  type PathsD,
  type Rect64,
  type RectD,
  ClipType,
  PathType,
  FillRule,
  PointInPolygonResult,
  type ZCallback64,
  type ZCallbackD,
  InternalClipper,
  Point64Utils,
  PointDUtils,
  Rect64Utils,
  RectDUtils,
  PathUtils,
  PathsUtils,
  InvalidRect64,
  InvalidRectD
} from './Core.js';

// Export engine classes
export {
  VertexFlags,
  Vertex,
  LocalMinima,
  createLocalMinima, // deprecated: use new LocalMinima() directly
  type IntersectNode,
  createIntersectNode,
  OutPt,
  JoinWith,
  HorzPosition,
  OutRec,
  HorzSegment,
  HorzJoin,
  Active,
  ClipperEngine,
  ReuseableDataContainer64,
  PolyPathBase,
  PolyPath64,
  PolyPathD,
  PolyTree64,
  PolyTreeD,
  ClipperBase,
  Clipper64,
  ClipperD
} from './Engine.js';

// Export offset functionality
export {
  JoinType,
  EndType,
  ClipperOffset
} from './Offset.js';

// Export rect clipping
export {
  OutPt2,
  RectClip64,
  RectClipLines64
} from './RectClip.js';

// Export Minkowski operations
export {
  Minkowski
} from './Minkowski.js';

// Export triangulation functionality
export {
  TriangulateResult,
  Delaunay
} from './Triangulation.js';

// Export main Clipper namespace with convenience functions
export * as Clipper from './Clipper.js';

// Re-export main functions for convenience
export {
  intersect,
  intersectD,
  union,
  unionD,
  difference,
  differenceD,
  xor,
  xorD,
  booleanOp,
  booleanOpWithPolyTree,
  booleanOpD,
  booleanOpDWithPolyTree,
  inflatePaths,
  inflatePathsD,
  rectClip,
  rectClipLines,
  minkowskiSum,
  minkowskiSumD,
  minkowskiDiff,
  minkowskiDiffD,
  area,
  areaPaths,
  areaD,
  areaPathsD,
  isPositive,
  isPositiveD,
  getBounds,
  getBoundsPaths,
  getBoundsD,
  getBoundsPathsD,
  makePath,
  makePathD,
  scalePath64,
  scalePaths64,
  scalePathD,
  scalePathsD,
  translatePath,
  translatePaths,
  translatePathD,
  translatePathsD,
  reversePath,
  reversePathD,
  reversePaths,
  reversePathsD,
  stripDuplicates,
  trimCollinear,
  trimCollinearD,
  pointInPolygon,
  pointInPolygonD,
  ellipse,
  ellipseD,
  simplifyPath,
  simplifyPaths,
  simplifyPathD,
  simplifyPathsD,
  ramerDouglasPeucker,
  ramerDouglasPeuckerPaths,
  ramerDouglasPeuckerD,
  ramerDouglasPeuckerPathsD,
  triangulate,
  triangulateD,
  polyTreeToPaths64,
  polyTreeToPathsD
} from './Clipper.js';
